sdk/scala-sdk/src/main/scala/kalix/scalasdk/action/Action.scala

```diff
@@ -181,6 +186,23 @@ abstract class Action {
     _actionContext = context
   }
 
+  /**
+   * Returns a [[TimerScheduler]] that can be used to schedule further in time.
+   */
+  final def timers: TimerScheduler = {
+
+    val javaActionContextImpl =
+      actionContext("Timers can only be scheduled or cancelled when handling a message.") match {
```

Comment:
Not sure if it is a huge problem, but, this will limit composability a bit, that you can't do something like:

```scala
val result = asyncTask().flatMap { result => 
  if (result.something) timers.startSingleTimer(some call)
  else Future.successful(whatever)
}
effects().asyncReply(result)
```

Maybe it makes more sense to create the `TimerSchedulerImpl` up front and pass through some internal field of the `ActionContext`? Or simply just always creating it on action creation as a final val.

Reply:
Creating it/injecting up front (maybe through `ActionContext`) would also make testkitting it easier.

Reply:
Good catch!

I think that injecting the client upfront should work. 

About composability, we will have the same issue with components. We should also break it up there. If possible.

Reply:
@johanandren, I think you have already faced that issue before and came out with a workaround.

See https://github.com/lightbend/kalix-jvm-sdk/blob/530c676ad904f40ac434642fdbc19df9994291d9/sdk/java-sdk/src/main/scala/kalix/javasdk/impl/action/ActionRouter.scala#L160-L162

We can't have it as a val in Action. Action is an open interface that anyone can implement and we need to have control over it as we need the actor system and the any support at hand. 

We could add the `timers()` method to the `ActionCreationContext`, but I think it's convenient to have it as a method in Action, like `Action.effects()` because it make it discoverable and because users will interact with timers and effects, so makes sense to keep them next to each other. 

Anyway, I think your idea of only setting it once works. I guess you added it for components, no?

Reply:
was this fixed or still remaining?

Reply:
I think we can leave it be as is and revisit refactoring/pulling it into the ActionContextImpl to be able to provide a pretend timer in the testkit.

Reply:
As far as I understand, the point originally raised is not about being able to replace the timer for the testkit, but if it's safe to access the actionContext in a map/flatMap/etc. And it's safe for the same reason it's safe to to call `components()` in a map/flatMap.

We can create the TimerScheduler beforehand and set it in the ActionContext, but the fact that we can only access it inside a command handler remains.